### PR TITLE
PR: Improve the display of tracebacks and better handle namespace and __file__ during execution

### PR DESCRIPTION
**Pull request by impact27**
This is the companion of PR spyder-ide/spyder#10047.

----

# Before:

## runfile:

before

## runcell:

before_runcell

# After:

## runfile:

after

Less internal levels

## runcell:

after_runcell
The file name is correct.

### CODE REVIEW

**Comment by impact27** [comment]
I don't think these errors are related to my work:
```error
_________ ERROR collecting spyder_kernels/utils/tests/test_nsview.py __________
spyder_kernels\utils\tests\test_nsview.py:38: in <module>
    PANEL = pd.Panel({0: pd.DataFrame([1,2]), 1:pd.DataFrame([3,4])})
E   TypeError: object() takes no parameters
```

**Comment by ccordoba12** [comment]
I think this one can be closed now that you merged it with #112, right?

**Comment by impact27** [comment]
This can be merged after #112

**Comment by ccordoba12** [comment]
@impact27, please clean the git history of this PR.

**Comment by impact27** [comment]
The runfile function has this line: `ipython_shell.user_ns.update(namespace)`. Is that something that should also be run with runcell?

**Comment by ccordoba12** [comment]
> The runfile function has this line: ipython_shell.user_ns.update(namespace). Is that something that should also be run with runcell?

@bcolsen?

**Comment by bcolsen** [comment]
>The runfile function has this line: `ipython_shell.user_ns.update(namespace)`. Is that something that should also be run with runcell?

That's only necessary to deal with the merging the new namespace that the file was run in with the current console namespace. Runcell is always in the current console namespace.

With the new namespace class I think that `ipython_shell.user_ns.update(namespace)` would be best in the `__exit__` method there. It only needs to be run if `current_namespace` is `False`.

**Comment by bcolsen** [comment]
What namespace is `debugfile` run in currently?

File: spyder_kernels/customize/spydercustomize.py [hunk 1]
```diff
@@ -1007,26 +1052,25 @@ def runcell(cellname, filename=None):
...
-    # Restore __file__
-    if namespace_has_file:
-        namespace['__file__'] = namespace_filename
-    else:
-        namespace.pop('__file__')
+    with FileNamespace(filename, current_namespace=True) as namespace:
+        try:
+            if PY2:
+                filename = maybe_encode(filename)
+            exec(compile(cell_code, filename, 'exec'), namespace)
```
Comment (bcolsen) [inline spyder_kernels/customize/spydercustomize.py]: If we don't use IPython's `run_cell`, does that mean that IPython magics still work?

Comment (impact27) [reply spyder_kernels/customize/spydercustomize.py]: Do we expect them to work? I mean if I run a file they don't work, wo why should they work when running a cell?

Comment (impact27) [reply spyder_kernels/customize/spydercustomize.py]: In my opinion, cells are just file fragments

Comment (bcolsen) [reply spyder_kernels/customize/spydercustomize.py]: It would be nice. We are trying to make it so you can run a file as cells like a Jupyter notebook.

**Comment by impact27** [comment]
So should runfile also accept magic?

File: spyder_kernels/customize/spydercustomize.py (hunk 1 above)
Comment (bcolsen) [reply spyder_kernels/customize/spydercustomize.py]: Cell magic current works for cells in spyder3

**Comment by bcolsen** [comment]
> So should runfile also accept magic?

The IPython `%run` magic only allows magics in .ipy and .ipynb they just run all the cells in a notebook file or the whole file in the case of an .ipy file. We could support .ipy file with spyder style cells or try to upstream that idea to IPython. We could also just have a button in spyder to run all the cells. 

https://github.com/ipython/ipython/blob/8971c8e5f31d20768609e20f60870d627eb33bd0/IPython/core/interactiveshell.py#L2741

It looks like they also can't get the file name for the trace backs but give a little warning at the end saying file they were running when the errors happened.

If we have to make a choice between having magics(I think IPython `run_cell` also does some async stuff and module reloads) or having filenames in tracebacks, I would pick magics and the stability of IPython `run_cell`(my opinion). 

In general IPython functions get a lot of testing and we should use then when we can. I know that this is working around problem with the debugger, but the closer we stay to upstream then the lighter the bug fixing load there is here. There are only so many corner cases we can test.

**Comment by impact27** [comment]
The main problem I see is that if the name of the file is not set properly, you can not use breakpoints with Pdb. An alternative would be to parse the cell for magic and call IPython explicitly.

File: spyder_kernels/customize/spydercustomize.py (hunk 1 above)
Comment (bcolsen) [reply spyder_kernels/customize/spydercustomize.py]: It looks like vscode can runcells with magics

**Comment by bcolsen** [comment]
> An alternative would be to parse the cell for magic and call IPython explicitly.

IPython has `.run_cell_magic` and  `.run_line_magic` so that might work

**Comment by impact27** [comment]
`ipython_shell.run_cell` was removed in #112, so I am not sure this is in scope of this PR. I think this should be discussed in a separate issue.

File: spyder_kernels/customize/spydercustomize.py [hunk 2]
```diff
@@ -867,8 +867,62 @@ def get_debugger(filename):
...
+        Reset namespace['__file__']
+        """
+        if not self.current_namespace:
+            get_ipython().user_ns.update(self.namespace)
+        if self._previous_filename:
+            self.namespace['__file__'] = self._previous_filename
+        elif '__file__' in self.namespace:
+            # Avoid error when running `%reset -f` programmatically
+            # See issue spyder-ide/spyder-kernels#91
+            self.namespace.pop('__file__')
```
Comment (ccordoba12) [inline spyder_kernels/customize/spydercustomize.py]: Please enclose this in a try/except KeyError (as before) to avoid the issue referenced above this line.

File: spyder_kernels/customize/spydercustomize.py [hunk 3]
```diff
@@ -867,8 +867,62 @@ def get_debugger(filename):
     return debugger, filename
 
 
+class FileNamespace():
```
[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder_kernels/customize/spydercustomize.py]: Also, please make this class inherit from `object` to be a new class in Python 2.

File: spyder_kernels/customize/spydercustomize.py [hunk 4]
```diff
@@ -898,52 +952,43 @@ def runfile(filename=None, args=None, wdir=None, namespace=None,
...
+                ipython_shell.run_cell_magic('cython', '', f.read())
+        else:
+            try:
+                execfile(filename, namespace)
+            except SystemExit as status:
+                # ignore exit(0)
+                if status.code:
+                    ipython_shell.showtraceback(exception_only=True)
+            except bdb.BdbQuit:
+                if not is_pdb:
```
Comment (ccordoba12) [inline spyder_kernels/customize/spydercustomize.py]: Please add a comment above this line to describe what you're doing here.

File: spyder_kernels/customize/spydercustomize.py [hunk 5]
```diff
@@ -898,52 +952,43 @@ def runfile(filename=None, args=None, wdir=None, namespace=None,
...
+                    ipython_shell.showtraceback(tb_offset=2)
+            except BaseException:
+                # tb offset is 1 because we wrap execfile
```
Comment (ccordoba12) [inline spyder_kernels/customize/spydercustomize.py]: Could you better describe here what part of the traceback we're discarding here?

File: spyder_kernels/customize/spydercustomize.py [hunk 6]
```diff
@@ -1007,26 +1053,25 @@ def runcell(cellname, filename=None):
...
+        try:
+            if PY2:
+                filename = maybe_encode(filename)
+            exec(compile(cell_code, filename, 'exec'), namespace)
+        except SystemExit as status:
+            # ignore exit(0)
+            if status.code:
+                ipython_shell.showtraceback(exception_only=True)
+        except bdb.BdbQuit:
+            if not is_pdb:
```
Comment (ccordoba12) [inline spyder_kernels/customize/spydercustomize.py]: Please add the same comments here as above.

File: spyder_kernels/customize/spydercustomize.py (hunk 1 above)
Comment (ccordoba12) [reply spyder_kernels/customize/spydercustomize.py]: I don't think we should allow magics in Python files because they're not valid code.

File: spyder_kernels/customize/spydercustomize.py (hunk 2 above)
Comment (impact27) [reply spyder_kernels/customize/spydercustomize.py]: Isn't `if '__file__' in self.namespace` enough? I thought the `KeyError` was to catch that.

File: spyder_kernels/customize/spydercustomize.py (hunk 5 above)
Comment (impact27) [reply spyder_kernels/customize/spydercustomize.py]: Sure, we discard the call to `execfile` and the call to `exec`

File: spyder_kernels/customize/spydercustomize.py (hunk 2 above)
Comment (ccordoba12) [reply spyder_kernels/customize/spydercustomize.py]: `%reset -f` removes it after `runfile` is executed, so this could still happen (I haven't tested it with this PR though). Also, since you left the comment above, I thought it still happened.

File: spyder_kernels/customize/spydercustomize.py (hunk 5 above)
Comment (ccordoba12) [reply spyder_kernels/customize/spydercustomize.py]: Ok, then please that explanation instead of the comment in this line.

File: spyder_kernels/customize/spydercustomize.py (hunk 2 above)
Comment (impact27) [reply spyder_kernels/customize/spydercustomize.py]: I left the comment to explain the `if '__file__' in self.namespace`. I tried spyder-ide/spyder-kernels#91 and it works as expected

File: spyder_kernels/customize/namespace_manager.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        """
+        Reset the namespace.
+        """
+        if not self.current_namespace:
+            get_ipython().user_ns.update(self.namespace)
+        if self._previous_filename:
+            self.namespace['__file__'] = self._previous_filename
+        elif '__file__' in self.namespace:
+            # Avoid error when running `%reset -f` programmatically
+            # See issue spyder-ide/spyder-kernels#91
```
Comment (ccordoba12) [inline spyder_kernels/customize/namespace_manager.py]: The two comments above no longer apply after your changes in this PR, so please remove them.
